### PR TITLE
Add registry

### PR DESCRIPTION
**Pull request by pierallard**
# What what what ?
- Change deprecated ProgressHelper by ProgressBar
- Add messages to ProgressBar to show what's generating
- Set a unique Generator to generate fixtures and products (old FixtureGenerator becomes an EntityGenerator)
- Automatically creates missing directory of fixtures
- Rename Generators with right names (AttributeGroupsAccesses becomes AttributeGroupAccesses, etc)
- Add a registry to register Generators. The EntityGenerator is more readable, and adding a generator is more simple.
- Rename `generate` params to help next developers
- Adds Specs
- And other things.

### CODE REVIEW

File: Command/AbstractGenerateCommand.php [hunk 1]
```diff
@@ -70,7 +103,12 @@ protected function checkOutputDirExists($outputDir)
     {
         $fs = new Filesystem();
         if (!$fs->exists($outputDir)) {
-            throw new \RuntimeException(sprintf('The output directory "%s" does not exist!', $outputDir));
+            mkdir($outputDir, 0777, true);
```
Comment (arnolanglade) [inline Command/AbstractGenerateCommand.php:106]: :+1:

File: Command/AbstractGenerateCommand.php [hunk 2]
```diff
@@ -20,26 +23,56 @@
...
         return $processedConfig;
     }
 
     /**
+     * {@inheritdoc}
+     */
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $configFile = $input->getArgument('configuration-file');
+
```
Comment (aRn0D) [inline Command/AbstractGenerateCommand.php]: you should remove the 3 next blank lines

File: DependencyInjection/RegisterGeneratorsPass.php [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * Compiler pass that register generators
+ *
+ * @author    Pierre Allard <pierre.allard@akeneo.com>
+ * @copyright 2016 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class RegisterGeneratorsPass implements CompilerPassInterface
```
Comment (arnolanglade) [inline DependencyInjection/RegisterGeneratorsPass.php]: Missing spec

File: Generator/AssetCategoryAccessGenerator.php [hunk 4]
```diff
@@ -29,9 +30,9 @@ public function __construct(CsvWriter $writer)
     /**
      * {@inheritdoc}
      */
-    public function generate(array $globalConfig, array $config, ProgressHelper $progress, array $options = [])
+    public function generate(array $globalConfig, array $entitiesConfig, ProgressBar $progress, array $options = [])
     {
-        $groups             = $options['groups'];
+        $groups             = $options['user_groups'];
```
Comment (aRn0D) [inline Generator/AssetCategoryAccessGenerator.php]: useless indentation

File: Generator/AssetCategoryAccessGenerator.php [hunk 5]
```diff
@@ -59,5 +60,15 @@ public function generate(array $globalConfig, array $config, ProgressHelper $pro
                 self::ASSET_CATEGORY_ACCESSES_FILENAME
             ))
             ->write($data);
+
+        return [];
```
Comment (aRn0D) [inline Generator/AssetCategoryAccessGenerator.php]: Why ?

File: Generator/EntitiesGenerator.php [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+use Symfony\Component\Console\Helper\ProgressBar;
+
+/**
+ * Fixture generator that will dispatch generation to specialized generator
+ *
+ * @author    Pierre Allard <pierre.allard@akeneo.com>
+ * @copyright 2016 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class EntitiesGenerator
```
Comment (aRn0D) [inline Generator/EntitiesGenerator.php]: Missing spec !!!!!!!!!!

File: Generator/GeneratorRegistry.php [hunk 7]
```diff
@@ -0,0 +1,42 @@
+<?php
+
+namespace Pim\Bundle\DataGeneratorBundle\Generator;
+
+/**
+ * @author    Pierre Allard <pierre.allard@akeneo.com>
+ * @copyright 2016 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class GeneratorRegistry
```
Comment (aRn0D) [inline Generator/GeneratorRegistry.php]: Spec

Comment (arnolanglade) [inline Generator/GeneratorRegistry.php]: Interface ?

File: Generator/VariantGroupGenerator.php [hunk 8]
```diff
@@ -265,7 +267,7 @@ protected function setAttributes(array $attributes)
     protected function setGroupTypes(array $groupTypes)
     {
         foreach ($groupTypes as $groupType) {
-            if ($groupType->getCode() == 'VARIANT') {
+            if ($groupType->getCode() === 'VARIANT') {
```
Comment (arnolanglade) [inline Generator/VariantGroupGenerator.php:270]: Constant ?

File: Resources/config/services.yml [hunk 9]
```diff
@@ -12,3 +12,6 @@ services:
             - '@pim_data_generator.product.product_value_raw_builder'
             - '@pim_catalog.repository.attribute'
             - '@pim_catalog.repository.category'
+
+    pim_data_generator.generator.registry:
+        class: 'Pim\Bundle\DataGeneratorBundle\Generator\GeneratorRegistry'
```
Comment (arnolanglade) [inline Resources/config/services.yml]: parameter ?

File: Resources/config/generators.yml [hunk 10]
```diff
@@ -14,148 +15,173 @@ parameters:
...
-            - @pim_data_generator.generator.group_type
-            - @pim_data_generator.generator.variant_group
-            - @pim_data_generator.generator.locale
+    pim_data_generator.entities_generator:
+        class: %pim_data_generator.generator.entities.class%
+        arguments:
+            - '@pim_data_generator.generator.registry'
 
     pim_data_generator.generator.channel:
         class: %pim_data_generator.generator.channel.class%
```
Comment (arnolanglade) [inline Resources/config/generators.yml]: parameter should be quoted

File: Generator/VariantGroupGenerator.php (hunk 8 above)
Comment (pierallard) [inline Generator/VariantGroupGenerator.php:270]: There is no "VARIANT" constant in the PIM :/

Comment (arnolanglade) [inline Generator/VariantGroupGenerator.php:270]: yoda style

**Pull request closed without merging**